packages/core/src/sync-realtime/service.ts

```diff
@@ -202,444 +219,506 @@ export class RealtimeSyncService extends Emittery<RealtimeSyncEvents> {
+    toBlockNumber: bigint,
+    insertChildAddress: boolean,
+  ): Promise<RealtimeLog[]> => {
+    if (this.hasFactorySource) {
+      // Find and insert any new child contracts.
+      const matchedFactoryLogs = filterLogs({
+        logs,
+        logFilters: this.factorySources.map((fs) => ({
+          address: fs.criteria.address,
+          topics: [fs.criteria.eventSelector],
+        })),
       });
 
-      if (!parentBlock_)
-        throw new Error(
-          `Failed to fetch parent block with hash: ${canonicalBlock.parentHash}`,
-        );
+      if (insertChildAddress) {
+        await this.syncStore.insertFactoryChildAddressLogs({
+          chainId: this.network.chainId,
+          logs: matchedFactoryLogs,
+        });
+      }
 
-      canonicalBlocksWithTransactions.unshift(
-        parentBlock_ as BlockWithTransactions,
+      // Find any logs matching log filters or child contract filters.
+      // NOTE: It might make sense to just insert all logs rather than introduce
+      // a potentially slow DB operation here. It's a tradeoff between sync
+      // latency and database growth.
+      // NOTE: Also makes sense to hold factoryChildAddresses in memory rather than
+      // a query each interval.
+      const factoryLogFilters = await Promise.all(
+        this.factorySources.map(async (factory) => {
+          const iterator = this.syncStore.getFactoryChildAddresses({
+            chainId: this.network.chainId,
+            factory: factory.criteria,
+            upToBlockNumber: toBlockNumber,
+          });
+          const childContractAddresses: Address[] = [];
+          for await (const batch of iterator) {
+            childContractAddresses.push(...batch);
+          }
+          return {
+            address: childContractAddresses,
+            topics: factory.criteria.topics,
+          };
+        }),
       );
-      depth += 1;
-      canonicalBlock = rpcBlockToLightBlock(parentBlock_);
 
-      this.common.logger.warn({
-        service: "realtime",
-        msg: `Fetched canonical block at height ${canonicalBlock.number} while reconciling reorg (network=${this.network.name})`,
+      return filterLogs({
+        logs,
+        logFilters: [
+          ...this.logFilterSources.map((l) => l.criteria),
+          ...factoryLogFilters,
+        ],
+      });
+    } else {
+      return filterLogs({
+        logs,
+        logFilters: this.sources.map((s) => s.criteria),
       });
     }
+  };
 
-    // 5) If the common ancestor was not found in our local chain, this is a deep reorg.
-    this.emit("deepReorg", {
-      detectedAtBlockNumber: newBlock.number,
-      minimumDepth: depth,
-    });
-
-    this.common.logger.warn({
-      service: "realtime",
-      msg: `Unable to reconcile >${depth}-block reorg (network=${this.network.name})`,
-    });
+  /** Returns true if "blocks" is not missing any blocks from tip to "finalizedBlockNumber". */
+  private isBlocksComplete = (): boolean => {
```

Comment:
`isLocalChainConsistent` or something similar? Was not sure what this meant when reading it above.

Reply:
Yeah this is a way better name